Panel content reorganization + Control/Tool rework

MichaelBuessemeyer opened the pull request:
### URL of deployed dev instance (used for testing):
- https://panelcontentreorganization.webknossos.xyz

### Steps to test:
- Look at the checklist below and check whether the panels/tabs have the content as described in the list.
- Please also check the (basic) functionality of the panel elements / settings.

- I had quite a lot of merging conflicts relating the dark mode pr / css. This needs to be tested, so nothing breaks because of this PR.

### Issues:
- fixes #5231

------
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- ~~[ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable~~
- ~~[ ] Updated [documentation](../blob/master/docs) if applicable~~
- ~~[ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes~~
- ~~[ ] Needs datastore update after deployment~~
- [x] Ready for review

MichaelBuessemeyer commented:
TODO List:

 Tabs
- [x] Layers (left)
  - Layers with histograms
  - Including segmentation layer or volume annotation layer
     - With mapping setting
- [ ] remove highlight hovered cells
- [ ] Settings (left)
  - Controls
  - Viewport Options (with "Display planes in 3D", without "Clipping Distance")
  - Data Rendering
- [ ] Info
    - As is (+ Dataset edit button)
- [ ] Skeleton
    - Nodes & Trees settings (including "Clipping distance")
- [ ] Trees
    - As is
 - [ ] Comments
    - As is
- [ ] Abstract Tree
    - As is
- [ ] Volume
    - Volume Options
    - unlink segmentation layer
- [ ] Meshes
   - As is
- [ ] BBoxes
   - Bounding Boxes from left sidebar
- [ ] Statusbar or Context Menu
  - Current Segment ID (under cursor)
  - Currently rendered mag

MichaelBuessemeyer commented:
This PR should be ready for the first review. Please note the open TODOs in the list are due to missing / not merged PRs.

@philippotto Could you do the review here because you were also part of the discussion meeting for this PR? If you do not have the time for that, feel free to ask / assign someone else 😄.

philippotto commented:
Thanks for the reorganization! Already looks quite good :+1: I shared this internally to gather more feedback, since I don't want to decide this fully on my own :)

MichaelBuessemeyer commented:
We discussed the following changes:

- Remove Skeleton and Volume-Tab again. The settings will be put somewhere else.
- Add a "Skeleton"- Layer. This layer also contains a switch at the top left corner to hide all settings and automatically hide all trees. The settings: Node Radius, Particle Size, Clipping Distance, Override Node Radius, Highlight Commented Nodes, Center new Nodes will then be listed under that layer
- Soma-clicking and Merger Mode should become tools in the Navbar (see #5403)
- Add Active Node/Tree Id to the status bar. If they are clicked the user can edit them via a popover.
- Do the same with the Active Cell-Id (Volume).
- Also put the brush size into the navbar (see #5403)
- Add ID at mouse position to the context menu to be able to copy it.
- Remove the info "ID at the center"
- Do not have two tabset in the right sidebar/boder. Put them all into one container without a vertical split as default. 
- Rename the "general" tab to something like "Rendering/Controls" or so.
- The "Layers" tab should be first in the left sidebar/border
- The ID Mapping Setting should be a direct part of the Segmentation/Volume Annotation Layer settings. Thus the collapses in this tab can be removed

philippotto commented:
@MichaelBuessemeyer I'd recommend to not do the context-menu-related items (at least now), as I'm working on that code area in my current PRs. Just so you know :)

MichaelBuessemeyer commented on frontend/stylesheets/main.less:
```diff
@@ -456,41 +457,3 @@ img.img-50 {
 .link-in-progress {
   cursor: wait;
 }
-
-.keyboard-key-icon {
-  border: 2px solid #888;
-  border-radius: 2px;
-  min-width: 26px;
-  min-height: 26px;
-  margin: 4px;
-  display: inline-block;
-  text-align: center;
-  font-size: 10px;
-  vertical-align: middle;
-  line-height: 24px;
-  font-weight: bold;
-}
-
-.keyboard-key-icon-small {
-  border: 1px solid rgb(255, 255, 255);
-  border-radius: 2px;
-  min-width: 14px;
```
This css is only used in the tracing view. Thus I moved it to the match less file

MichaelBuessemeyer commented:
The PR should be ready for a first feedback session @philippotto. Please also see my comment in #5069.

MichaelBuessemeyer commented:
Open Feedback:

 - [ ] Make the buttons in the skeletons and the comments tab tighter to better fit in one row.
 - [ ] Let the toggle of the skeleton layer not toggle the tree visibility of each tree individually but let it hide all trees by hiding the threejs group (or so) of the skeletons.
 - [ ] Make Soma Clicking and Merger Mode Button Icons monochrome
 - [ ] Get rid of Cell Id in the UI and name it "segment Id" everywhere
 - [ ] Improve BBox "+" Button
     - Put in the same line as the text
     - Use the same button as used for isosurfaces
     - Maybe make it primary as it is the onliest button.
- [ ] Redesign a BBox entry. e.g. BBox name in the same line as the switch/toggle button and only make it an editable text field if the user clicked a pen icon to the right.
- [ ] Creating a new annotation: The skeleton layer is initially deactivated, which feels weird. This should be resolved when the functionality of the skeleton layer activation/deactivation is changed.
- [ ] Investigate: Pattern opacity is only visible in volume annotation, but not view mode (intentional?) 
- [ ] 


TODO Later when @philippotto's PRs are ready:
 - [ ] Make sure to hide the "Change brush size"-button in skeleton only tracings.
 - [ ] Add the "merger mode" and the "soma clicking" button to the other buttons.
 - [ ] As Volume Mode is not active together with merger mode, merger mode should become an own mode like skeleton and brush, trace, fill, and cell picker.
 - [ ] Reorder the toolbar buttons.

Open for discussion: 
 - Better Icons for the merger mode and soma clicking. 
    - Soma clicking / merger mode icons are very eye-catching compared to the others, not sure if we want that.

philippotto commented on frontend/javascripts/oxalis/view/statusbar.js:
```diff
@@ -205,30 +199,44 @@ class Statusbar extends React.PureComponent<Props, State> {
     );
   }
 
+  maybeLabelWitSegmentationWarning = (label: string) =>
```
typo in "with"

philippotto reviewed (commented):
Whew, quite a lot of changes. To be frank, I only did high-level code review since so many lines were changed that I couldn't check everything line-by-line.

I left some small comments, but nothing major. After you incorporated my feedback, I suggest to merge this PR together with #5422. Afterwards, we can do at least one testing round and gather final feedback. Merging the two PRs will probably be a bit painful due to merge conflicts. Maybe it makes sense to have a call and do it together? Would tomorrow (Wednesday) work for you?

philippotto commented:
pulling the master into the branch might make more sense **after** we merged the two prs together. let's discuss this together in person. so, no need to get started with resolving merge conflicts already :)

philippotto commented:
Some further feedback I have now after testing (**but let's incorporate this later since this would produce additional merge conflicts otherwise):

- the tabs "Layer" and "Controls/Rendering" might need a better name
	- maybe "Layers" and "Settings" ?
- the editable UI components in the statusbar are not recognizable as such
	- definitely, change the mouse cursor when hovering
	- maybe a small edit icon would help? might take too much space, though
- the new buttons in the toolbar could be more context-sensitive (e.g., only show the brush size UI if the brush is active)
- styling
  - the warning icon here looks a bit weirdly placed:
    ![image](https://user-images.githubusercontent.com/2486553/120339145-26489880-c2f5-11eb-9b41-7786851dbf5b.png)
  - the margins look weird (ideally, these should be consistent with the comments tab)
![image](https://user-images.githubusercontent.com/2486553/120339224-36607800-c2f5-11eb-90f0-8120e7dd0b16.png)

philippotto commented:
I'm starting a merge of this PR with #5422 on another branch. Wish me luck :sweat_smile:

MichaelBuessemeyer commented on frontend/javascripts/oxalis/model/reducers/skeletontracing_reducer_helpers.js:
```diff
@@ -746,9 +746,11 @@ export function deleteComment(
 export function toggleAllTreesReducer(
   state: OxalisState,
   skeletonTracing: SkeletonTracing,
+  setVisible?: boolean,
```
thanks for finding this :)

This was previously needed in a state of this PR, where deactivating the "skeleton-layer" actually toggled the visibility prop of each tree. Now that this behaviour is changed, this is indeed not needed anymore :+1:

MichaelBuessemeyer commented on frontend/javascripts/oxalis/constants.js:
```diff
@@ -96,24 +96,19 @@ export const OrthoViewCrosshairColors: OrthoViewMap<[number, number]> = {
 
-export const TracingTabs = {
+export const BorderTabs = {
+  // TODO: Somehow add a tooltip and option to have a short name to the title
```
This is no longer a todo, as I already used abbreviation names all the time.

philippotto commented:
I just pushed the merged version onto this branch. I didn't do much testing, yet, but superficially everything looked okay. I'll do more testing as a next step and gather todos. However, I will not commit anything anymore until we decide otherwise. This should give you space, @MichaelBuessemeyer, to tackle your remaining todos without running into conflicts. I hope this makes sense!

MichaelBuessemeyer commented:
@philippotto I don't remember whether I notified you, but I think I applied all feedback from last friday. Could you please take another look and tell me, whether this state is ok give to users to ask for feedback?
Thanks :) 🙏

philippotto commented on frontend/javascripts/oxalis/constants.js:
```diff
@@ -97,17 +97,46 @@ export const OrthoViewCrosshairColors: OrthoViewMap<[number, number]> = {
+  AbstractTreeTab: {
+    id: "AbstractTreeTab",
+    name: "AbsTree",
+    description: "Views an abstract version of the active skeleton",
```
```suggestion
    description: "Shows an abstract version of the active skeleton",
```

I think, "to view" would be the "act of looking" (which is done by the user and not be the tab).

philippotto commented on frontend/javascripts/oxalis/view/layouting/flex_layout_wrapper.js:
```diff
@@ -250,6 +250,11 @@ class FlexLayoutWrapper extends React.PureComponent<Props, State> {
         model={model}
         factory={(...args) => this.layoutFactory(...args)}
+        titleFactory={renderedNode => (
+          <Tooltip title={BorderTabs[renderedNode.getId()].description}>
```
For abbreviated tab titles, the tooltip should ideally show "Long Name - Description". For example, hovering over "AbsTree" should show a tooltip with "Abstract Tree - Shows ...".

philippotto commented on frontend/javascripts/oxalis/view/action-bar/toolbar_view.js:
```diff
@@ -156,6 +170,49 @@ function OverwriteModeSwitch({ isControlPressed, isShiftPressed, visible }) {
+  const toggleNewNodeNewTreeMode = () =>
+    dispatch(updateUserSettingAction("newNodeNewTree", !isNewNodeNewTreeModeOn));
+  const toggleMergerMode = () => dispatch(setMergerModeEnabledAction(!isMergerModeEnabled));
+  const activeButtonStyle = { ...narrowButtonStyle, borderColor: "#1890ff" };
```
Did you check the border color in both themes? Ideally, this would use an antd variable, such as `--ant-box-shadow-base`.

MichaelBuessemeyer replied:
good catch :)

philippotto commented on frontend/javascripts/oxalis/view/action-bar/toolbar_view.js:
```diff
@@ -156,6 +170,49 @@ function OverwriteModeSwitch({ isControlPressed, isShiftPressed, visible }) {
+
+  return (
+    <React.Fragment>
+      <Tooltip title="Toggle the Single node Tree (soma clicking) mode">
```
I know that has been 

```suggestion
      <Tooltip title="Toggle the Single node Tree (soma clicking) mode – If enabled, each node creation will create a new tree.">
```

philippotto commented on frontend/javascripts/oxalis/view/action-bar/toolbar_view.js:
```diff
@@ -156,6 +170,49 @@ function OverwriteModeSwitch({ isControlPressed, isShiftPressed, visible }) {
+          />
+        </ButtonComponent>
+      </Tooltip>
+      <Tooltip title="Toggle Merger Mode">
```
```suggestion
      <Tooltip title="Toggle Merger Mode – When enabled, trees which connect multiple segments will merge those segments.">
```

philippotto reviewed (commented):
Awesome :) only left some nitpicks

philippotto commented:
- [ ] superfluous divider in context menu when opening a dataset with a segmentation in view mode
- [ ] when soma-clicking AND merger mode are enabled, the border look weird (there is still some radius between this buttons?)
- [ ] create-new-tree icon needs to be changed (@philippotto)
- [ ] "Layer" to "Layers"
- [ ] edit icon next to "active segment id" etc. could be a bit more subtle and have better padding (@philippotto)
- [ ] padding-top of skeleton-tab still seems to be too big

- [ ] "not all layers visible" still looks a bit weird. also, the icon pushes everything further down which makes the the other switch-buttons hard to use

To discuss (let's discuss this after we got some feedback):
- [ ] the "volume" tab is gone now. still not sure, whether people will easily cope.

philippotto commented:
I think I pushed a solution for all items I mentioned above. Feel free to review my latest commits. I used the existing `antd-legacy-group` CSS class and removed the one you added, as the existing class seems to work better regarding the removal of the border-radius. I hope, I didn't oversee any regressions.

hotzenklotz commented on frontend/javascripts/oxalis/view/novel_user_experiences/01-present-modern-controls.js:
```diff
@@ -0,0 +1,88 @@
+    <Modal maskClosable={false} visible onCancel={closeModal} width={800} footer={null}>
+      <h1>Say Hello to the Context Menu</h1>
+      <p>
+        webKnossos now provides an easy-to-use context menu which allows to perform even complex
```
```suggestion
        webKnossos now provides an easy-to-use context menu that allows performing even complex
```

hotzenklotz commented on frontend/javascripts/oxalis/view/novel_user_experiences/01-present-modern-controls.js:
```diff
@@ -0,0 +1,88 @@
+      <h1>Say Hello to the Context Menu</h1>
+      <p>
+        webKnossos now provides an easy-to-use context menu which allows to perform even complex
+        actions, such as merging two trees, in an intuitive way. Simply rightclick an element of
```
```suggestion
        actions, such as merging two trees, intuitively. Simply right-click an element of
```

hotzenklotz commented on frontend/javascripts/oxalis/view/novel_user_experiences/01-present-modern-controls.js:
```diff
@@ -0,0 +1,88 @@
+      />
+      <p>
+        Previously, the right-click was reserved for certain actions, such as creating a node or
+        erasing volume data. These actions are now available in their respective tool via leftclick
```
```suggestion
        erasing volume data. These actions are now available in their respective tool via left-click
```

hotzenklotz commented on frontend/javascripts/oxalis/view/novel_user_experiences/01-present-modern-controls.js:
```diff
@@ -0,0 +1,88 @@
+      <p>
+        Previously, the right-click was reserved for certain actions, such as creating a node or
+        erasing volume data. These actions are now available in their respective tool via leftclick
+        (e.g., using the dedicated erase-tool) or happen automatically (e.g., selecting a node
```
```suggestion
        (e.g., using the dedicated erase tool) or happen automatically (e.g., selecting a node
```

hotzenklotz commented on docs/skeleton_annotation.md:
```diff
@@ -98,9 +109,9 @@ Hover over existing tree groups to bring up a little menu for creating new group
-  1. Select a node at which to split (*SHIFT + Left Click*) and delete (*DEL*) it.
-  2. Delete an edge between two nodes. Select the first node (*SHIFT + Left Click*), then *SHIFT + CTRL + Left Click* on the second node of the edge to delete this connection.
-- `Tree merging` works similarly to edge deletion but will create a new edge between two previously unconnected trees. Select the first node and *SHIFT + ALT + Left Click* the second node to create an edge between the two.
+  1. Delete a node at which to split. This can be done by rightclicking a node and choosing "Delete this Node". If you have enabled *Classic Controls*, you need to select (*SHIFT + Left Click*) the node first and then delete (*DEL*) it.
```
```suggestion
  1. Delete a node at which to split. This can be done by right-clicking a node and choosing "Delete this Node". If you have enabled *Classic Controls*, you need to select (*SHIFT + Left Click*) the node first and then delete (*DEL*) it.
```

hotzenklotz commented on docs/skeleton_annotation.md:
```diff
@@ -98,9 +109,9 @@ Hover over existing tree groups to bring up a little menu for creating new group
-  2. Delete an edge between two nodes. Select the first node (*SHIFT + Left Click*), then *SHIFT + CTRL + Left Click* on the second node of the edge to delete this connection.
-- `Tree merging` works similarly to edge deletion but will create a new edge between two previously unconnected trees. Select the first node and *SHIFT + ALT + Left Click* the second node to create an edge between the two.
+  1. Delete a node at which to split. This can be done by rightclicking a node and choosing "Delete this Node". If you have enabled *Classic Controls*, you need to select (*SHIFT + Left Click*) the node first and then delete (*DEL*) it.
+  2. Delete an edge between two nodes. Select the first node (*Left Click*), then rightclick the second node and select *Delete Edge to this Node*. If you have enabled *Classic Controls*, you need to select the first node with *Shift + Left Click* and then click on the second node with *SHIFT + CTRL + Left Click* on the second node of the edge to delete this connection.
```
```suggestion
  2. Delete an edge between two nodes. Select the first node (*Left Click*), then right-click the second node and select *Delete Edge to this Node*. If you have enabled *Classic Controls*, you need to select the first node with *Shift + Left Click* and then click on the second node with *SHIFT + CTRL + Left Click* on the second node of the edge to delete this connection.
```

hotzenklotz commented on docs/skeleton_annotation.md:
```diff
@@ -98,9 +109,9 @@ Hover over existing tree groups to bring up a little menu for creating new group
-- `Tree merging` works similarly to edge deletion but will create a new edge between two previously unconnected trees. Select the first node and *SHIFT + ALT + Left Click* the second node to create an edge between the two.
+  1. Delete a node at which to split. This can be done by rightclicking a node and choosing "Delete this Node". If you have enabled *Classic Controls*, you need to select (*SHIFT + Left Click*) the node first and then delete (*DEL*) it.
+  2. Delete an edge between two nodes. Select the first node (*Left Click*), then rightclick the second node and select *Delete Edge to this Node*. If you have enabled *Classic Controls*, you need to select the first node with *Shift + Left Click* and then click on the second node with *SHIFT + CTRL + Left Click* on the second node of the edge to delete this connection.
+- `Tree merging` works similarly to edge deletion but will create a new edge between two previously unconnected trees. Select the first node and rightclick the second to choose *Create Edge & Merge with this Tree*. When using *Classic Controls*, the second node needs to be selected with *SHIFT + ALT + Left Click* to create an edge between the two.
```
```suggestion
- `Tree merging` works similarly to edge deletion but will create a new edge between two previously unconnected trees. Select the first node and right-click on a second one to choose *Create Edge & Merge with this Tree*. When using *Classic Controls*, the second node needs to be selected with *SHIFT + ALT + Left Click* to create an edge between the two.
```

hotzenklotz commented on docs/skeleton_annotation.md:
```diff
@@ -73,14 +73,25 @@ Press the "J" to jump to the latest branch point to continue working from there
 | J                 | Jump to Last Branch Point           |
 | C                 | Create New Tree                     |
 
+Note that you can enable *Classic Controls* which will behave slightly different and more explicit for the mouse actions:
```
```suggestion
Note that you can enable *Classic Controls* which will behave slightly different and more explicit for some mouse actions:
```

hotzenklotz commented on docs/keyboard_shortcuts.md:
```diff
@@ -77,24 +99,28 @@ Find all available keyboard & mouse shortcuts for webKnossos listed below.
 | V                                 | Copy Segmentation of Current Cell From Previous Slice       |
 | SHIFT + V                         | Copy Segmentation of Current Cell From Next Slice           |
 
+Note that you can enable *Classic Controls* which won't open a context-menu on rightclick, but instead erases when the brush/trace tool is activated.
```
```suggestion
Note that you can enable *Classic Controls* which won't open a context-menu on right-click, but instead erases when the brush/trace tool is activated.
```

hotzenklotz commented on docs/keyboard_shortcuts.md:
```diff
@@ -1,6 +1,15 @@
 # Keyboard & Mouse Shortcuts
 
-Find all available keyboard & mouse shortcuts for webKnossos listed below.
+The most important shortcuts are always shown in the statusbar at the bottom.
```
```suggestion
The most important shortcuts are always shown in the statusbar at the bottom of the screen.
```

hotzenklotz commented on docs/keyboard_shortcuts.md:
```diff
@@ -1,6 +1,15 @@
+
+Note that you can enable *Classic Controls* in the left sidebar.
+Without classic controls (which is the default), a more intuitive behavior is used which assigns the most important functionality to the left mouse button (e.g., moving around, selecting/creating/moving nodes) while the right mouse button always opens a context-sensitive menu for more complex actions, such as merging two trees.
+With classic controls, several mouse controls are modifier-driven and may also use the rightclick for actions, such as erasing volume data.
```
```suggestion
With classic controls, several mouse controls are modifier-driven and may also use the right-click for actions, such as erasing volume data.
```

hotzenklotz commented on docs/getting_started.md:
```diff
@@ -50,11 +50,18 @@ You can also change the size of the viewports to see more details in your data a
-Basic movement in the dataset is done with the mouse wheel or by pressing the spacebar keyboard shortcut.
+webKnossos will launch the main annotation screen allowing you to navigate your dataset, place markers to reconstruct skeletons, or annotate segments as volume annotations.
+
+Depending on the current tool which can be selected in the top bar, various actions can be performed.
```
```suggestion
Depending on the current tool - selectable in the top bar - various actions can be performed.
```

hotzenklotz commented on docs/getting_started.md:
```diff
@@ -50,11 +50,18 @@ You can also change the size of the viewports to see more details in your data a
+
+Depending on the current tool which can be selected in the top bar, various actions can be performed.
+Note that the most important controls are always shown in the status bar at the bottom of your screen.
+The first tool is the move tool which allows to navigate the dataset by moving the mouse while holding the left mouse button.
```
```suggestion
The first tool is the MOVE tool which allows to navigate the dataset by moving the mouse while holding the left mouse button.
```

hotzenklotz commented on docs/getting_started.md:
```diff
@@ -50,11 +50,18 @@ You can also change the size of the viewports to see more details in your data a
+Depending on the current tool which can be selected in the top bar, various actions can be performed.
+Note that the most important controls are always shown in the status bar at the bottom of your screen.
+The first tool is the move tool which allows to navigate the dataset by moving the mouse while holding the left mouse button.
+In the skeleton tool, left click can be used to place markers in the data, called nodes.
```
```suggestion
In the SKELETON tool, left click can be used to place markers in the data, called nodes.
```

hotzenklotz commented on docs/getting_started.md:
```diff
@@ -50,11 +50,18 @@ You can also change the size of the viewports to see more details in your data a
+The first tool is the move tool which allows to navigate the dataset by moving the mouse while holding the left mouse button.
+In the skeleton tool, left click can be used to place markers in the data, called nodes.
+Additionally, the left mouse button can also be used to navigate around, select or drag nodes.
+The draw tool allows to "paint" voxels to create volumetric annotations.
```
```suggestion
The DRAW tool allows to "paint" voxels to create volumetric annotations.
```

hotzenklotz reviewed (commented):
I read through many of the documentation changes and modals texts. In some cases I doubled checked with Grammarly.

MichaelBuessemeyer commented:
> I think I pushed a solution for all items I mentioned above. Feel free to review my latest commits. I used the existing antd-legacy-group CSS class and removed the one you added, as the existing class seems to work better regarding the removal of the border-radius. I hope, I didn't oversee any regressions.

Thanks for doing that :). I took a look at your changes and they look alright to me :)
I added a small fix, that if only the soma clicking mode is active, the border on the right side of the button was not shown / not blue by simply adding a `z-index: 1` to the button if it is active. At least this solved the issue for me.
